src/modules/flow/led-7seg/led-7seg.c

```diff
@@ -0,0 +1,265 @@
+{
+    int i;
+
+    if (!mdata->common_cathode)
+        byte = ~byte;
+
+    for (i = 0; i < 8; i++) {
+        bool val = (byte >> i) & 1;
+        if (!sol_gpio_write(mdata->gpio[i], val)) {
+            SOL_WRN("Failed to write on gpio %" PRId32 ".", i);
+            return -EIO;
+        }
+    }
+
+    return 0;
+}
+
+static int
+segments_set(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    int r;
+    unsigned char in_value;
+
+    r = sol_flow_packet_get_byte(packet, &in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = write_byte(data, node, in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+value_set(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    int32_t in_value;
+    int r;
+
+#define RANGE_MIN (0)
+#define RANGE_MAX (15)
+
+    r = sol_flow_packet_get_irange_value(packet, &in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    if ((in_value < RANGE_MIN) || (in_value > RANGE_MAX)) {
+        sol_flow_send_error_packet(node, ERANGE,
+            "Range invalid, it should be between %d and %d but was %d",
+            RANGE_MIN, RANGE_MAX, in_value);
+        return 0;
+    }
+
+    r = write_byte(data, node, font[in_value]);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+
+#undef RANGE_MAX
+#undef RANGE_MIN
+}
+
+#define OPEN_GPIO(_pin, _option) \
+    do { \
+        mdata->gpio[_pin] = sol_gpio_open(opts->pin_ ## _option.val, \
+            &gpio_conf); \
+        if (!mdata->gpio[_pin]) { \
+            SOL_WRN("could not open gpio #%" PRId32, \
+                opts->pin_ ## _option.val); \
+            return -EIO; \
+        } \
+    } while (0)
+
+static int
+led_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    struct led_7seg_data *mdata = data;
+    const struct sol_flow_node_type_led_7seg_led_options *opts =
+        (const struct sol_flow_node_type_led_7seg_led_options *)options;
+    struct sol_gpio_config gpio_conf = { 0 };
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options,
+        SOL_FLOW_NODE_TYPE_LED_7SEG_LED_OPTIONS_API_VERSION, -EINVAL);
+
+    gpio_conf.api_version = SOL_GPIO_CONFIG_API_VERSION;
+    gpio_conf.dir = SOL_GPIO_DIR_OUT;
+
+    OPEN_GPIO(0, dp);
+    OPEN_GPIO(1, g);
+    OPEN_GPIO(2, f);
+    OPEN_GPIO(3, e);
+    OPEN_GPIO(4, d);
+    OPEN_GPIO(5, c);
+    OPEN_GPIO(6, b);
+    OPEN_GPIO(7, a);
+
+    mdata->common_cathode = opts->common_cathode;
+
+    return 0;
+}
+
+#undef OPEN_GPIO
+
+static void
+led_close(struct sol_flow_node *node, void *data)
+{
+    struct led_7seg_data *mdata = data;
+    int i;
+
+    for (i = 0; i < 8; i++)
+        sol_gpio_close(mdata->gpio[i]);
+}
+
+/* Conversion table between chars and bytes in format 'abcdefgX' */
+static unsigned char conversion_table[][2] = {
+    { '\0', 0x0 },
```

Comment:
Maybe initialize this as:

    [0] = 0x00,
    [' '] = 0x00,
    ['0'] = 0xfc,
    ...

Instead of having to loop through this array? (This is also a more compact representation.)